File: crates/hm-dsl-engine/harmont-ts/src/keygen.ts
import { createHash } from "node:crypto";
import {
  readFileSync,
  readdirSync,
  statSync,
  existsSync,
} from "node:fs";
import { join, resolve as resolvePath, relative } from "node:path";
import type { PipelineIR } from "./pipeline.js";

const NUL = "\0";

export interface CacheKeyOptions {
  readonly pipelineOrg: string;
  readonly pipelineSlug: string;
  readonly now: number;
  readonly basePath: string;
  readonly env: Readonly<Record<string, string>>;
}

export function resolvePipelineCacheKeys(
  graph: PipelineIR["graph"],
  opts: CacheKeyOptions,
): void {
  const nodes = graph.nodes;
  const edges = graph.edges;

  const keyByIdx = new Map<number, string>();
  for (let i = 0; i < nodes.length; i++) {
    keyByIdx.set(i, nodes[i].step.key as string);
  }

  const parentKeyMap = new Map<string, string>();
  for (const [src, dst, kind] of edges) {
    if (kind === "builds_in") {
      parentKeyMap.set(keyByIdx.get(dst)!, keyByIdx.get(src)!);
    }
  }

  const resolved = new Map<string, string>();

  for (const node of nodes) {
    const step = node.step;
    const cache = step.cache as Record<string, unknown> | undefined;
    if (!cache || cache.policy === "none") continue;

    const cmd = (step.cmd as string) ?? "";
    const stepKey = step.key as string;
    const parentStepKey = parentKeyMap.get(stepKey);
    const parentResolved = lookupParent(parentStepKey, resolved);
    const policyRes = resolvePolicy(cache, cmd, opts);

    const key = sha256hex(
      opts.pipelineOrg +
        NUL +
        opts.pipelineSlug +
        NUL +
        stepKey +
        NUL +
        parentResolved +
        NUL +
        policyRes,
    );

    cache.key = key;
    resolved.set(stepKey, key);
  }
}

function lookupParent(
  parentStepKey: string | undefined,
  resolved: Map<string, string>,
): string {
  if (parentStepKey == null) return "scratch";
  const key = resolved.get(parentStepKey);
  if (key == null) {
    throw new Error(
      `step references builds_in "${parentStepKey}" which has no cached key (parent must be defined upstream and cached)`,
    );
  }
  return key;
}

function resolvePolicy(
  cache: Record<string, unknown>,
  cmd: string,
  opts: CacheKeyOptions,
): string {
  const policy = cache.policy as string;

  if (policy === "none") return "none";

  if (policy === "forever") {
    const envKeys = (cache.env_keys as string[]) ?? [];
    return "forever-" + sha256hex(cmd + NUL + envSubset(envKeys, opts.env));
  }

  if (policy === "ttl") {
    const duration = cache.duration_seconds as number;
    const bucket = Math.floor(opts.now / duration);
    const envKeys = (cache.env_keys as string[]) ?? [];
    return (
      "ttl-" +
      bucket +
      "-" +
      sha256hex(cmd + NUL + envSubset(envKeys, opts.env))
    );
  }

  if (policy === "on_change") {
    const paths = (cache.paths as string[]) ?? [];
    const resolvedPaths: string[] = [];
    for (const p of [...paths].sort()) {
      if (/[*?[]/.test(p)) {
        resolvedPaths.push(...globPaths(opts.basePath, p));
      } else {
        const full = resolvePath(opts.basePath, p);
        if (existsSync(full)) resolvedPaths.push(full);
      }
    }
    const pre = resolvedPaths.map((r) => pathHash(r) + NUL).join("");
    return "sha-" + sha256hex(pre);
  }

  if (policy === "compose") {
    const subs = cache.sub_policies as Record<string, unknown>[];
    const parts = subs.map((sub) =>
      sub.policy !== "none" ? resolvePolicy(sub, cmd, opts) : "none",
    );
    return "compose-" + sha256hex(parts.join(""));
  }

  throw new Error(`resolve-policy-key: unknown policy "${policy}"`);
}

function envSubset(
  envKeys: readonly string[],
  env: Readonly<Record<string, string>>,
): string {
  const sorted = [...envKeys].sort();
  return sorted.map((k) => k + "=" + (env[k] ?? "") + NUL).join("");
}

function pathHash(fullPath: string): string {
  const stat = statSync(fullPath);
  if (stat.isFile()) {
    return sha256hex(readFileSync(fullPath));
  }
  if (stat.isDirectory()) {
    const h = createHash("sha256");
    const files = walkDir(fullPath).sort();
    for (const child of files) {
      const rel = relative(fullPath, child).split("\\").join("/");
      h.update(rel, "utf8");
      h.update(NUL);
      h.update(readFileSync(child));
      h.update(NUL);
    }
    return h.digest("hex");
  }
  throw new Error(`on_change path does not exist: ${fullPath}`);
}

function walkDir(dir: string): string[] {
  const results: string[] = [];
  for (const entry of readdirSync(dir, { withFileTypes: true })) {
    const full = join(dir, entry.name);
    if (entry.isDirectory()) {
      results.push(...walkDir(full));
    } else if (entry.isFile()) {
      results.push(full);
    }
  }
  return results;
}

function globPaths(basePath: string, pattern: string): string[] {
  const parts = pattern.split("/");
  let candidates = [basePath];

  for (const part of parts) {
    const next: string[] = [];
    for (const dir of candidates) {
      if (!existsSync(dir) || !statSync(dir).isDirectory()) continue;
      if (part === "**") {
        next.push(dir);
        next.push(...walkDir(dir).map((f) => join(f, "..")));
      } else if (/[*?[]/.test(part)) {
        const re = globPartToRegex(part);
        for (const entry of readdirSync(dir, { withFileTypes: true })) {
          if (re.test(entry.name)) {
            next.push(join(dir, entry.name));
          }
        }
      } else {
        const full = join(dir, part);
        if (existsSync(full)) next.push(full);
      }
    }
    candidates = next;
  }

  return candidates
    .filter((p) => existsSync(p) && statSync(p).isFile())
    .sort();
}

function globPartToRegex(part: string): RegExp {
  let re = "^";
  for (const ch of part) {
    if (ch === "*") re += ".*";
    else if (ch === "?") re += ".";
    else re += ch.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  }
  re += "$";
  return new RegExp(re);
}

function sha256hex(data: string | Buffer): string {
  return createHash("sha256").update(data).digest("hex");
}

File: crates/hm-dsl-engine/src/keygen.rs
//! Cache-key resolver — byte-exact port of `harmont-py/harmont/keygen.py`.
//!
//! The output bytes MUST match the Python (and, before it, Scheme) resolver so
//! cache snapshots persisted by earlier versions stay reachable. The outer key
//! is the sha256 of the preimage
//!
//! ```text
//! pipeline_org NUL pipeline_slug NUL step_key NUL parent_resolved NUL policy_resolution
//! ```
//!
//! where `NUL` is a single `\x00` byte and `policy_resolution` branches on the
//! step's cache policy (see [`resolve_policy`]).

use std::collections::BTreeMap;
use std::fmt::Write as _;
use std::path::{Path, PathBuf};

use anyhow::{Context, Result, bail};
use sha2::{Digest, Sha256};

use crate::step_chain::RawCachePolicy;

/// The `\x00` separator woven through every preimage.
const NUL: &str = "\x00";

/// Parameters cache-key resolution needs beyond the raw step chain.
///
/// Carried through the lowering pass so keys can be resolved while the rich
/// [`RawCachePolicy`] data (env keys, durations, paths, sub-policies) is still
/// in scope, before it is flattened into the IR `Cache` shape.
#[derive(Debug, Clone)]
pub struct LowerOptions {
    /// Owning organization slug — first field of the outer preimage.
    pub pipeline_org: String,
    /// Pipeline slug — second field of the outer preimage.
    pub pipeline_slug: String,
    /// Wall-clock unix timestamp used to bucket `ttl` policies.
    pub now: u64,
    /// Project directory that `on_change` paths and globs resolve against.
    pub base_path: PathBuf,
    /// Process environment sampled for `env_subset` of `forever`/`ttl` policies.
    pub env: BTreeMap<String, String>,
}

/// Hex-encode `sha256(s)` as lowercase, matching `hashlib.sha256(...).hexdigest()`.
#[must_use]
pub fn sha256_hex(s: &str) -> String {
    to_hex(&Sha256::digest(s.as_bytes()))
}

/// Resolve a step's full outer cache key.
///
/// `parent_resolved` is the already-resolved key of the `builds_in` parent, or
/// `"scratch"` when the step has no cached parent.
///
/// # Errors
///
/// Propagates errors from [`resolve_policy`] — chiefly missing `on_change`
/// paths or filesystem read failures.
pub fn compute_cache_key(
    step_key: &str,
    cmd: &str,
    policy: &RawCachePolicy,
    parent_resolved: &str,
    opts: &LowerOptions,
) -> Result<String> {
    let policy_res = resolve_policy(policy, cmd, opts)?;
    let preimage = [
        opts.pipeline_org.as_str(),
        opts.pipeline_slug.as_str(),
        step_key,
        parent_resolved,
        policy_res.as_str(),
    ]
    .join(NUL);
    Ok(sha256_hex(&preimage))
}

/// Resolve the `policy_resolution` fragment for a single policy.
///
/// # Errors
///
/// Returns an error if an `on_change` path is missing (and not silently
/// skippable) or a file/directory cannot be read.
pub fn resolve_policy(policy: &RawCachePolicy, cmd: &str, opts: &LowerOptions) -> Result<String> {
    match policy {
        RawCachePolicy::None => Ok("none".to_owned()),
        RawCachePolicy::Forever { env_keys } => {
            let inner = [cmd, &env_subset(env_keys, &opts.env)].join(NUL);
            Ok(format!("forever-{}", sha256_hex(&inner)))
        }
        RawCachePolicy::Ttl {
            duration_seconds,
            env_keys,
        } => {
            let bucket = opts.now / duration_seconds;
            let inner = [cmd, &env_subset(env_keys, &opts.env)].join(NUL);
            Ok(format!("ttl-{bucket}-{}", sha256_hex(&inner)))
        }
        RawCachePolicy::OnChange { paths } => {
            Ok(format!("sha-{}", sha256_hex(&on_change_preimage(paths, &opts.base_path)?)))
        }
        RawCachePolicy::Compose { sub_policies } => {
            let mut parts = String::new();
            for sub in sub_policies {
                if matches!(sub, RawCachePolicy::None) {
                    parts.push_str("none");
                } else {
                    parts.push_str(&resolve_policy(sub, cmd, opts)?);
                }
            }
            Ok(format!("compose-{}", sha256_hex(&parts)))
        }
    }
}

/// Concatenate `key=value\x00` for each sorted env key, reading `value` from
/// `env` (empty string when absent).
#[must_use]
pub fn env_subset(env_keys: &[String], env: &BTreeMap<String, String>) -> String {
    let mut sorted: Vec<&String> = env_keys.iter().collect();
    sorted.sort();
    let mut out = String::new();
    for k in sorted {
        out.push_str(k);
        out.push('=');
        out.push_str(env.get(k).map_or("", String::as_str));
        out.push_str(NUL);
    }
    out
}

/// Build the `on_change` preimage: `file_hash(p) NUL` for each resolved path.
///
/// Path strings are sorted first; glob patterns (`*`, `?`, `[`) expand against
/// `base` and their matches are sorted; plain paths that don't exist are
/// skipped (mirroring `keygen.py`).
fn on_change_preimage(paths: &[String], base: &Path) -> Result<String> {
    let mut sorted: Vec<&String> = paths.iter().collect();
    sorted.sort();

    let mut resolved: Vec<PathBuf> = Vec::new();
    for p in sorted {
        if p.contains('*') || p.contains('?') || p.contains('[') {
            let pattern = base.join(p);
            let pattern = pattern.to_str().with_context(|| {
                format!("on_change glob pattern is not valid UTF-8: {}", pattern.display())
            })?;
            let mut matches: Vec<PathBuf> = glob::glob(pattern)
                .with_context(|| format!("invalid on_change glob pattern: {pattern}"))?
                .collect::<std::result::Result<Vec<_>, _>>()
                .with_context(|| format!("failed to read on_change glob matches for {pattern}"))?;
            matches.sort();
            resolved.extend(matches);
        } else {
            let full = base.join(p);
            if full.exists() {
                resolved.push(full);
            }
        }
    }

    let mut pre = String::new();
    for r in &resolved {
        pre.push_str(&path_hash(r)?);
        pre.push_str(NUL);
    }
    Ok(pre)
}

/// Hash a path for an `on_change` key.
///
/// Files hash their bytes. Directories fold each descendant file's POSIX
/// relative path + bytes into one stream in sorted order. Missing paths fail
/// loudly.
fn path_hash(path: &Path) -> Result<String> {
    if path.is_file() {
        let bytes =
            std::fs::read(path).with_context(|| format!("reading on_change file {}", path.display()))?;
        return Ok(to_hex(&Sha256::digest(&bytes)));
    }
    if path.is_dir() {
        let mut files: Vec<PathBuf> = Vec::new();
        collect_files(path, &mut files)?;
        files.sort();

        let mut h = Sha256::new();
        for child in &files {
            let rel = child
                .strip_prefix(path)
                .with_context(|| format!("relativizing {}", child.display()))?;
            let rel_posix = rel
                .components()
                .map(|c| c.as_os_str().to_string_lossy())
                .collect::<Vec<_>>()
                .join("/");
            h.update(rel_posix.as_bytes());
            h.update([0u8]);
            let bytes = std::fs::read(child)
                .with_context(|| format!("reading on_change file {}", child.display()))?;
            h.update(&bytes);
            h.update([0u8]);
        }
        return Ok(to_hex(&h.finalize()));
    }
    bail!("on_change path does not exist: {}", path.display());
}

/// Collect every regular file under `dir`, recursing into subdirectories.
fn collect_files(dir: &Path, out: &mut Vec<PathBuf>) -> Result<()> {
    let entries =
        std::fs::read_dir(dir).with_context(|| format!("reading directory {}", dir.display()))?;
    for entry in entries {
        let path = entry
            .with_context(|| format!("reading entry in {}", dir.display()))?
            .path();
        if path.is_dir() {
            collect_files(&path, out)?;
        } else if path.is_file() {
            out.push(path);
        }
    }
    Ok(())
}

/// Lowercase hex encoding of a byte slice.
fn to_hex(bytes: &[u8]) -> String {
    let mut out = String::with_capacity(bytes.len() * 2);
    for b in bytes {
        let _ = write!(out, "{b:02x}");
    }
    out
}
